### PR TITLE
Statically evaluate mathematical binary expressions

### LINKED ISSUE #1085: Variables are not statically evaluated
**CSS classes are preferred over CSS variables since the former are statically evaluated.** 

However, there are instances where CSS variables are created. 
See examples below:

 

CSS variables are generated in the instance the interpolated variable comes as the result of a function:
```
const si = () => 8 // When the number comes as the result of a func
const gridSize = si();

const LayoutRight = styled.aside`
  margin-right: -${gridSize * 5}px;
  margin-left: ${gridSize * 5}px;
  top: -${gridSize * 5}px;
  left: -${gridSize * 8}px;
  right: ${gridSize * 8}px;
  color: red;
`;

// Generates ouput
const _6 = "._syaz5scu{color:red}";
const _5 = "._1xi21iyg{right:var(--_12ygqxf)}";
const _4 = "._1ltv1u8g{left:var(--_1prhge8)}";
const _3 = "._154isxb8{top:var(--_1cakqv5)}";
const _2 = "._18u01xn1{margin-left:var(--_1l3fmvo)}";
const _ = "._2hwxsxb8{margin-right:var(--_1cakqv5)}";
```

But [no function calculating size]
```
// const si = () => 8. // Disabled
const gridSize = 8;

const LayoutRight = styled.aside`
  margin-right: -${gridSize * 5}px;
  margin-left: ${gridSize * 5}px;
  top: -${gridSize * 5}px;
  left: -${gridSize * 8}px;
  right: ${gridSize * 8}px;
  color: red;
`;

// Generates output
const _6 = "._syaz5scu{color:red}";
const _5 = "._1xi21fwp{right:64px}";
const _4 = "._1ltv1vci{left:-64px}";
const _3 = "._154i10j7{top:-40px}";
const _2 = "._18u01ylp{margin-left:40px}";
const _ = "._2hwx10j7{margin-right:-40px}";
```

### PR DESCRIPTION
**Pull request by at-nathan**
Resolves #1085 

Adds ability to statically evaluate expressions like:
```
const spacing = 8;
<div css={{ padding: spacing * 2 }} />
```

### CODE REVIEW

File: packages/babel-plugin/src/utils/evaluate-expression.tsx [hunk 1]
```diff
@@ -89,7 +90,7 @@ const babelEvaluateExpression = (
     }
 
     const result = path.evaluate();
-    if (result.value) {
+    if (result.value != null) {
```
Comment (at-nathan) [inline packages/babel-plugin/src/utils/evaluate-expression.tsx:93]: Lets `0` be a valid value.

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 2]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
       '"--_12w6gfj":ix(getLineHeight())',
     ]);
   });
+
+  describe('Binary expresssions', () => {
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:398]:
```suggestion
  describe('binary expresssions', () => {
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 3]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+    it('statically evaluates calcuated value with identifier', () => {
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:399]:
```suggestion
    it('statically evaluates calculated value with identifier', () => {
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 4]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+      const actual = transform(`
+        import '@compiled/react';
+
+        const spacing = 8;
+
+        <div css={{marginTop: spacing * 2}} />
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:405]:
```suggestion
        <div css={{ marginTop: spacing * 2 }} />
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 5]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+      expect(actual).toIncludeMultiple(['._19pkexct{margin-top:16px}', 'ax(["_19pkexct"])']);
+    });
+
+    it('statically evaluates calcuated value with nested binary', () => {
+      const actual = transform(`
+        import '@compiled/react';
+
+        const spacing = 8;
+
+        <div css={{marginTop: spacing * 2 / 2}} />
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:417]:
```suggestion
        <div css={{ marginTop: spacing * 2 / 2 }} />
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 6]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+
+        const spacing = 8;
+
+        <div css={{marginTop: spacing * 2}} />
+      `);
+
+      expect(actual).toIncludeMultiple(['._19pkexct{margin-top:16px}', 'ax(["_19pkexct"])']);
+    });
+
+    it('statically evaluates calcuated value with nested binary', () => {
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:411]:
```suggestion
    it('statically evaluates calculated value with nested binary', () => {
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 7]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+
+        const spacing = 8;
+
+        <div css={{marginTop: spacing * 2 / 2}} />
+      `);
+
+      expect(actual).toIncludeMultiple(['._19pkftgi{margin-top:8px}', 'ax(["_19pkftgi"])']);
+    });
+
+    it('statically evaluates calcuated value with multiple identifiers', () => {
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:423]:
```suggestion
    it('statically evaluates calculated value with multiple identifiers', () => {
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 8]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+      const actual = transform(`
+        import '@compiled/react';
+
+        const one = 1;
+        const two = 2;
+        const three = 3;
+
+        <div css={{marginTop: one + two - three}} />
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:431]:
```suggestion
        <div css={{ marginTop: one + two - three }} />
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 9]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+      `);
+
+      expect(actual).toIncludeMultiple(['._19pkidpf{margin-top:0}', 'ax(["_19pkidpf"])']);
+    });
+
+    it('statically evaluates calcuated value within calc utility', () => {
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:437]:
```suggestion
    it('statically evaluates calculated value within calc utility', () => {
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 10]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+      const actual = transform(`
+        import '@compiled/react';
+
+        const spacing = 8;
+
+        <div css={{width: \`calc(100% - \${spacing * 2}px)\`}} />
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:443]:
```suggestion
        <div css={{ width: \`calc(100% - \${spacing * 2}px)\` }} />
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 11]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+      `);
+
+      expect(actual).toIncludeMultiple([
+        '._1bsbj0q6{width:calc(100% - 16px)}',
+        'ax(["_1bsbj0q6"])',
+      ]);
+    });
+
+    it('statically evaluates calcuated value with string literal containing numeric value', () => {
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:452]:
```suggestion
    it('statically evaluates calculated value with string literal containing numeric value', () => {
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 12]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+      const actual = transform(`
+        import '@compiled/react';
+
+        const stringSpacing = '8';
+
+        <div css={{marginTop: stringSpacing * 2}} />
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:458]:
```suggestion
        <div css={{ marginTop: stringSpacing * 2 }} />
```

File: packages/babel-plugin/src/__tests__/expression-evaluation.test.ts [hunk 13]
```diff
@@ -394,4 +394,87 @@ describe('import specifiers', () => {
...
+      expect(actual).toIncludeMultiple(['._19pkexct{margin-top:16px}', 'ax(["_19pkexct"])']);
+    });
+
+    it('falls back to dynamic evaluation when non static value used', () => {
+      const actual = transform(`
+        import '@compiled/react';
+
+        const getSpacing = () => Math.random();
+
+        <div css={{marginTop: getSpacing() * 2}} />
```
Comment (MonicaOlejniczak) [inline packages/babel-plugin/src/__tests__/expression-evaluation.test.ts:470]:
```suggestion
        <div css={{ marginTop: getSpacing() * 2 }} />
```

File: packages/babel-plugin/src/utils/traverse-expression/traverse-binary-expression.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+const hasNumericValue = (expression: t.Expression): boolean =>
+  t.isNumericLiteral(expression) ||
+  (t.isStringLiteral(expression) && !Number.isNaN(Number(expression.value)));
+
+export const traverseBinaryExpression = (
+  expression: t.BinaryExpression,
+  meta: Metadata,
+  evaluateExpression: EvaluateExpression
+): ReturnType<typeof createResultPair> => {
+  if (!t.isPrivateName(expression.left)) {
```
[the remark this answers is not in the document]
Comment (MonicaOlejniczak) [reply packages/babel-plugin/src/utils/traverse-expression/traverse-binary-expression.ts:20]: Does this produce an error if you leave it out?

[the remark this answers is not in the document]
Comment (at-nathan) [reply packages/babel-plugin/src/utils/traverse-expression/traverse-binary-expression.ts:20]: It doesn't. Not sure what a binary expression with a `PrivateName` looks like. But the `if` guard also doubles as type safety, as calculations should involve 2 numeric values. `PrivateName` to me sounds like `null` or `undefined`?

[the remark this answers is not in the document]
Comment (at-nathan) [reply packages/babel-plugin/src/utils/traverse-expression/traverse-binary-expression.ts:20]: hmm ok, so `PrivateName` are for private declarations in classes:
```
class Person {
  #name
}
```
I don't think they can be used in a calculation.

File: packages/babel-plugin/src/__tests__/css-builder.test.ts [hunk 15]
```diff
@@ -49,25 +49,4 @@ describe('css builder', () => {
       "
     `);
   });
-
```
Comment (at-nathan) [inline packages/babel-plugin/src/__tests__/css-builder.test.ts:52]: Moved to [behaviour.test.ts](https://github.com/atlassian-labs/compiled/pull/1139/files#diff-db85bb77fe1990a0e351b91b67d37bb7430d43fad07248e0da1192bb519eb115)

File: packages/babel-plugin/src/utils/css-builders.ts [hunk 16]
```diff
@@ -405,64 +405,6 @@ const extractLogicalExpression = (node: t.ArrowFunctionExpression, meta: Metadat
   return { css: mergeSubsequentUnconditionalCssItems(css), variables };
 };
 
-const createNegativeUnaryExpression = (node: t.Identifier) => ({
```
Comment (at-nathan) [inline packages/babel-plugin/src/utils/css-builders.ts:408]: Removed this solution as it was breaking for expressions like:
```
margin-right: -${(3 / 4) * gridSize}px;
```

File: packages/babel-plugin/src/utils/css-builders.ts [hunk 17]
```diff
@@ -701,9 +635,12 @@ const extractTemplateLiteral = (node: t.TemplateLiteral, meta: Metadata): CSSOut
     // E.g. `font-size: ${fontSize}px` will end up needing to look like:
     // `font-size: var(--_font-size)`, with the suffix moved to inline styles
     // style={{ '--_font-size': fontSize + 'px' }}
-    const name = `--_${hash(variableName)}`;
     const nextQuasis = node.quasis[index + 1];
     const [before, after] = cssAffixInterpolation(quasi.value.raw, nextQuasis.value.raw);
+    // Create a different CSS var name for negative value version
+    const name = `--_${hash(variableName)}${
```
Comment (at-nathan) [inline packages/babel-plugin/src/utils/css-builders.ts:641]: Replaced the negative number variable fix with this. Basically generate a different name for the negative version of the interpolation.

**Review by JakeLane — APPROVED** [review]

**Pull request merged**